nemo/collections/tts/models/spectrogram_enhancer.py

```diff
@@ -0,0 +1,382 @@
+from omegaconf import DictConfig, OmegaConf
+from pytorch_lightning import Trainer
+from pytorch_lightning.loggers import TensorBoardLogger, WandbLogger
+from torch import Tensor, nn
+from torch.autograd import grad as torch_grad
+from torch.utils.tensorboard.writer import SummaryWriter
+
+import nemo
+from nemo.collections.tts.models import FastPitchModel
+from nemo.collections.tts.models.base import SpectrogramGenerator
+from nemo.collections.tts.modules.spectrogram_enhancer import Discriminator, Generator, mask
+from nemo.core import Exportable, ModelPT
+from nemo.utils import logging
+
+
+def type_as_recursive(e, source: Tensor):
+    if isinstance(e, (list, tuple)):
+        return [type_as_recursive(elem, source) for elem in e]
+    elif isinstance(e, dict):
+        return {key: type_as_recursive(value, source) for key, value in e.items()}
+    elif isinstance(e, Tensor):
+        return e.type_as(source)
+    else:
+        return e
+
+
+def gradient_penalty(images, output, weight=10):
+    batch_size = images.shape[0]
+    gradients = torch_grad(
+        outputs=output,
+        inputs=images,
+        grad_outputs=torch.ones(output.size(), device=images.device),
+        create_graph=True,
+        retain_graph=True,
+        only_inputs=True,
+    )[0]
+
+    gradients = gradients.reshape(batch_size, -1)
+    return weight * ((gradients.norm(2, dim=1) - 1) ** 2).mean()
+
+
+def calc_pl_lengths(styles, images):
+    device = images.device
+    num_pixels = images.shape[2] * images.shape[3]
+    pl_noise = torch.randn(images.shape, device=device) / math.sqrt(num_pixels)
+    outputs = (images * pl_noise).sum()
+
+    pl_grads = torch_grad(
+        outputs=outputs,
+        inputs=styles,
+        grad_outputs=torch.ones(outputs.shape, device=device),
+        create_graph=True,
+        retain_graph=True,
+        only_inputs=True,
+    )[0]
+
+    return (pl_grads ** 2).sum(dim=2).mean(dim=1).sqrt()
+
+
+def gen_hinge_loss(fake):
+    return fake.mean()
+
+
+def hinge_loss(real, fake):
+    return (F.relu(1 + real) + F.relu(1 - fake)).mean()
+
+
+def consistency_loss(condition, output, lengths):
+    *_, w, h = condition.shape
+    w, h = w // 4, h
+
+    condition = torch.nn.functional.interpolate(condition, size=(w, h), mode="bilinear", antialias=True)
+    output = torch.nn.functional.interpolate(output, size=(w, h), mode="bilinear", antialias=True)
+
+    dist = (condition - output).abs()
+    dist = mask(dist, lengths)
+    return (dist / rearrange(lengths, "b 1 -> b 1 1 1")).sum(dim=-1).mean()
+
+
+class SpectrogramEnhancerModel(ModelPT, Exportable):
+    """GAN-based model to add details to blurry spectrograms from TTS models like Tacotron or FastPitch."""
+
+    def __init__(self, cfg: DictConfig, trainer: Trainer = None) -> None:
+        self.spectrogram_model = None
+        super().__init__(cfg=cfg, trainer=trainer)
+
+        self.init_spectrogram_model()
+
+        self.G = Generator(
```

Comment:
G, D needs to be instantiated from the yaml config.

Reply:
You mean passing a DictConfig to them? What is the rationale? Neither of them are ModelPT

Reply:
Not DictConfig object but providing the configs of your module in the yaml file like , I mean users should have the option to change G, D from the yaml file instead of touching the code. In the future, if someone wants to use a different Generator or Discriminator, it should be easier for them to just implement it and change in the yaml file instead of importing the modules in this file and changing the instantiation here. Doesn't really matter if they are inheriting ModelPT or not.

Example: duration predictor is instantiated here https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/tts/models/fastpitch.py#L143
And the corresponding config is provided here https://github.com/NVIDIA/NeMo/blob/main/examples/tts/conf/fastpitch_align_44100.yaml#L198
Now, if I want to use a different module as duration predictor I do not need to change the code, all I will need to do is to change the yaml file. Makes it easy and clean.

Reply:
> I mean users should have the option to change G, D from the yaml file instead of touching the code.

> In the future, if someone wants to use a different Generator or Discriminator, it should be easier for them to just implement it and change in the yaml file instead of importing the modules in this file and changing the instantiation here.

We use a specific config of a very specific generative model (StyleGAN 2), modified for working with sequences. Using your own generator or discriminator is discouraged.

Personally to me, this YAML-oriented design caused a lot of frustration when I was starting with NeMo and FastPitch. Some issues off the top of my head:
1) Readability. There is no class name in code, just a bare `instantiate` and a config field. It instantiates _something_, and one has to jump between the config and the code. (Half a year passed, and I'm still doing this)
2) Explorabilily. Which modules exactly can I put there? The ones with the name "Predictor" in it (FP example)?
3) Extensibility is illusionary. Want your new replacement submodule to take some extra input? Modify Python code, YAML can't do this. Or write a Python monster to make this YAML-configurable.

Overall, the "just change it in the yaml file, easy and clean" feature actually comes at cost of spiraling cognitive load when trying to understand the code. It also blurs the border between YAML configuration and Python programming, making both harder to work with.

While there are strong benefits for models like FastPitch, I don't think it is warranted here at all.

Reply:
Not exposing G and D in the YAML config was a conscious decision from my side. Unless it's an absolute requirement and blocks merging, I'd rather not do it.

Reply:
To not instantiate them with configurable parameters in the YAML would go against the NeMo design pattern. I do not know if it is the optimal approach, and its hard to say because in general we are clearly executing on it poorly (eg. 90% of the config files are copy-pasted making them way too large. NN layers which could be shared between models are instead copy-pasted between model definitions. The interfaces for the modules are not generic or documented).

But I think arguing that is out of the scope for a PR. Any individual PR should have to follow the coding styles and design patterns of the code base as decided by the organization.

Reply:
> would go against the NeMo design pattern

> But I think arguing that is out of the scope for a PR. Any individual PR should have to follow the coding styles and design patterns of the code base as decided by the organization.

OK, good points.

Reply:
Done